### PR TITLE
upload pycontw fb pages insights data

### PR DESCRIPTION
**Pull request by grimmer0125**
## Types of changes

- **New feature**

## Description

- Get facebook token first and fill them in `.env`, ref: https://github.com/pycontw/python-fb-page-insights-client#get-needed-secrets-first
- Manually trigger `fb_page_insights_data_uploader.py` or schedule a Airflow task to run (setup ready)
- Please check the below data on BigQuery which reflect same data shown in FB page insights portal 
    - [ods_pycontw_fb_page_summary_insights](https://console.cloud.google.com/bigquery?project=pycontw-225217&ws=!1m30!1m4!4m3!1spycontw-225217!2sods!3sods_sponsorQuestionnaire_ip_datetime!1m4!4m3!1spycontw-225217!2sods!3sods_questionnaire_ip_datetime!1m4!4m3!1spycontw-225217!2sods!3sods_pycontw_fb_page_summary_insights!1m4!4m3!1spycontw-225217!2sods!3sods_pycontw_fb_posts!1m4!4m3!1spycontw-225217!2sods!3sods_pycontw_fb_posts_insights!1m4!4m3!1spycontw-225217!2sods!3spycontw_fb_page_summary_insight&d=ods&p=pycontw-225217&t=ods_pycontw_fb_page_summary_insights&page=table)
    - [ods_pycontw_fb_posts_insights](https://console.cloud.google.com/bigquery?project=pycontw-225217&ws=!1m25!1m4!4m3!1spycontw-225217!2sods!3sods_sponsorQuestionnaire_ip_datetime!1m4!4m3!1spycontw-225217!2sods!3sods_questionnaire_ip_datetime!1m4!4m3!1spycontw-225217!2sods!3sods_pycontw_fb_page_summary_insights!1m4!4m3!1spycontw-225217!2sods!3sods_pycontw_fb_posts!1m4!4m3!1spycontw-225217!2sods!3sods_pycontw_fb_posts_insights&d=ods&p=pycontw-225217&t=ods_pycontw_fb_posts_insights&page=table) (period API parameter only can be lifetime, so we can not `directly` get a history data chart of a post via `1 api call`, the workaround way is to query a post's data periodically and plot something like `x1=lifetime_data_until_0708, x2= lifetime_data_until_0709` 
    - [ods_pycontw_fb_posts](https://console.cloud.google.com/bigquery?project=pycontw-225217&ws=!1m20!1m4!4m3!1spycontw-225217!2sods!3sods_sponsorQuestionnaire_ip_datetime!1m4!4m3!1spycontw-225217!2sods!3sods_questionnaire_ip_datetime!1m4!4m3!1spycontw-225217!2sods!3sods_pycontw_fb_page_summary_insights!1m4!4m3!1spycontw-225217!2sods!3sods_pycontw_fb_posts&d=ods&p=pycontw-225217&t=ods_pycontw_fb_posts&page=table)

Currently, the Airflow schedule_interval is to fetch the whole week page insights weekly & whole year posts daily. Can discuss it is suitable or not. 

## Checklist:
- [ ] Add test cases to all the changes you introduce
- [x] Run `poetry run pytest` locally to ensure all linter checks pass
- [ ] Update the documentation if necessary

### CODE REVIEW

File: contrib/fb_pages_insights/test.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+print("start")
```
Comment (tai271828) [inline contrib/fb_pages_insights/test.py:1]: shebang is suggested.

File: contrib/fb_pages_insights/test.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    }
+    pycontw_page_id = '160712400714277'
+    fb = FBPageInsight(**args)
+    data = fb.get_page_full(pycontw_page_id, until_date=(2020, 11, 15))
+
+def test_bigquery_client():
+    from google.cloud import bigquery
+
+    project=os.getenv("BIGQUERY_PROJECT")
+    #BIGQUERY_PROJECT=pycontw-225217
```
Comment (tai271828) [inline contrib/fb_pages_insights/test.py:24]: If the code is not used, maybe we should remove them.

File: pyproject.toml [hunk 3]
```diff
@@ -12,6 +12,7 @@ searchconsole = "^0.0.4"
 google-cloud-bigquery = "<2.8.0"
 pandas = "^1.2.3"
 pyarrow = "^3.0.0"
+python-fb-page-insights-client = {git = "git@github.com:pycontw/python-fb-page-insights-client.git", rev = "poetry3"}
```
Comment (tai271828) [inline pyproject.toml:15]: If we need another pyproject, maybe let us create another repository for this codebase?

**Review by tai271828 — COMMENTED** [review]
Not aware of this is WIP until now. I have left some comments for your reference. Thank you for the effort!

File: contrib/fb_page_insights/fb_page_insights_data_uploader.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+from pydantic import BaseSettings
+from typing import List, Dict, Union
+
+from google.cloud import bigquery
+from google.oauth2 import service_account
+
+
+class Settings(BaseSettings):
+    GOOGLE_APPLICATION_CREDENTIALS = ""
+    BIGQUERY_PROJECT = ""
```
Comment (david30907d) [inline contrib/fb_page_insights/fb_page_insights_data_uploader.py:19]: cool~

File: contrib/fb_page_insights/fb_page_insights_data_uploader.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    dataset_id = "ods"
+    settings = Settings()
+    credentials = service_account.Credentials.from_service_account_file(
+        settings.GOOGLE_APPLICATION_CREDENTIALS
+    )
+    # TODO: avoid duplicate instantiate twice
+    client = bigquery.Client(credentials=credentials, project=settings.BIGQUERY_PROJECT)
+
+    # convert json schema to bigquery schema
+    bigquery_schema: List[bigquery.SchemaField] = []
```
Comment (david30907d) [inline contrib/fb_page_insights/fb_page_insights_data_uploader.py:39]: 👍

File: dags/ods/fb_page_insights/dags/fb_page_insights_2_bigquery.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+"""
+Send PyconTW Facebook Page Insights Data to BigQuery
+"""
```
Comment (david30907d) [inline dags/ods/fb_page_insights/dags/fb_page_insights_2_bigquery.py:3]: nice 👍

File: dags/ods/fb_page_insights/dags/fb_page_insights_2_bigquery.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    "schedule_interval": "@weekly",
+    "retries": 2,
+    "retry_delay": timedelta(minutes=5),
+    "max_active_runs": 1,
+    "catchup": False,
+    "on_failure_callback": lambda x: "Need to send notification to Discord",
+}
+
+
+with DAG("FB_PAGE_INSIGHTS_2_BIGQUERY", default_args=DEFAULT_ARGS) as dag:
```
Comment (david30907d) [inline dags/ods/fb_page_insights/dags/fb_page_insights_2_bigquery.py:25]: cool, learned a new way to create `DAG` 😄

File: dags/ods/fb_page_insights/dags/fb_page_insights_2_bigquery.py (hunk 6 above)
Comment (david30907d) [reply dags/ods/fb_page_insights/dags/fb_page_insights_2_bigquery.py:3]: btw, @grimmer0125 if you get a chance, would you rename this file to `dag.py` plz ?

File: dags/ods/fb_page_insights/udfs/fb_page_insights_data_uploader.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    FBPageInsight,
+    PageWebInsightData,
+    PostsWebInsightData,
+    DatePreset,
+    Period,
+)
+from pydantic import BaseSettings
+from typing import List, Dict, Union
+
+from google.cloud import bigquery
```
Comment (david30907d) [inline dags/ods/fb_page_insights/udfs/fb_page_insights_data_uploader.py:13]: btw, I forgot to mention that we should use [bigquery hook](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/_api/airflow/providers/google/cloud/hooks/bigquery/index.html) or [bigquery operator](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/operators/cloud/bigquery.html). It's a better way to handle connections in Airflow

File: dags/ods/fb_page_insights/dags/fb_page_insights_2_bigquery.py (hunk 6 above)
Comment (grimmer0125) [reply dags/ods/fb_page_insights/dags/fb_page_insights_2_bigquery.py:3]: Renaming is easy. I'm curious that is any convention to use `dag.py` or any advantage? Generally speaking, multiple files using the same file name in a code base is a little hard to maintain and search.

File: dags/ods/fb_page_insights/udfs/fb_page_insights_data_uploader.py (hunk 8 above)
Comment (grimmer0125) [reply dags/ods/fb_page_insights/udfs/fb_page_insights_data_uploader.py:13]: Due to the running out of time and no other file in this repo already starts to use `bigquery hook` or `bigquery operator`, is it good to try these two in another PR?

**Comment by grimmer0125** [comment]
@tai271828 , @david30907d  CI failed on `ERROR tests/test_cakeresume_uploader.py`, not sure it is the same issue in https://github.com/pycontw/pycon-etl/issues/31

File: Dockerfile [hunk 9]
```diff
@@ -9,12 +9,15 @@ RUN apt-get update \
...
-    && pip install --no-cache-dir poetry==1.0.5 \
+    && pip install --no-cache-dir poetry==1.1.7 \
     && apt-get clean \
     && rm -rf /var/lib/apt/lists/*
 
 COPY pyproject.toml pyproject.toml
 COPY poetry.lock poetry.lock
+# RUN apt-get install -y python-setuptools 
+# RUN pip install -U pip
+# RUN pip install -U setuptools
```
Comment (david30907d) [inline Dockerfile:20]: do we need this?

File: Dockerfile [hunk 10]
```diff
@@ -9,12 +9,15 @@ RUN apt-get update \
     # 1. if you don't need postgres, remember to remove postgresql-dev and sqlalchemy
     # 2. libglib2.0-0 libsm6 libxext6 libxrender-dev libgl1-mesa-dev are required by opencv
     # 3. git is required by pip install git+https
-    && pip install --no-cache-dir poetry==1.0.5 \
+    && pip install --no-cache-dir poetry==1.1.7 \
```
Comment (david30907d) [inline Dockerfile:12]: 👍

File: dags/ods/fb_page_insights/dags/fb_page_insights_2_bigquery.py (hunk 6 above)
Comment (david30907d) [reply dags/ods/fb_page_insights/dags/fb_page_insights_2_bigquery.py:3]: yea you're right, it's just my convention. It's okay to keep this naming~

**Review by david30907d — APPROVED** [review]
utACK

File: Dockerfile (hunk 9 above)
Comment (grimmer0125) [reply Dockerfile:20]: No, it is commented when I do some experiments about poetry which has some issues on its older version. I will delete them.

**Comment by grimmer0125** [comment]
@david30907d  I pushed one more commit that diving the DAG to two DAGs, one is for page and the other one is for posts.

**Pull request merged**